Add tests for failed SSL connections: bad cert, and bad alpn config

ctiller opened the pull request:
Test suite tries to connect a client to a misconfigured server.
Servers can be substituted.

Built on #4442

jboeuf commented on test/core/bad_ssl/server.c:
```diff
@@ -0,0 +1,92 @@
+  int shutdown_finished = 0;
+  grpc_event ev;
+	
+	grpc_completion_queue *cq = grpc_completion_queue_create(NULL);
```
clang-format

jboeuf commented on test/core/bad_ssl/servers/alpn.c:
```diff
@@ -0,0 +1,81 @@
+  const char *addr = bad_ssl_addr(argc, argv);
+  grpc_ssl_pem_key_cert_pair pem_key_cert_pair = {test_server1_key,
+                                                  test_server1_cert};
+	grpc_server_credentials *ssl_creds;
```
clang-format

jboeuf commented on test/core/bad_ssl/servers/alpn.c:
```diff
@@ -0,0 +1,81 @@
+#include "test/core/bad_ssl/server.h"
+#include "test/core/end2end/data/ssl_test_data.h"
+
+static const char *const fake_versions[] = {"not-h2"};
```
Maybe have a comment explaining the scenario here. 
e.g. regular client will connect to a server advertising a bad alpn version (see fake_version).

jboeuf commented on test/core/bad_ssl/server.c:
```diff
@@ -0,0 +1,92 @@
+#include "test/core/bad_ssl/server.h"
+#include "test/core/util/test_config.h"
+
+static int got_sigint = 0;
```
Maybe a comment explaining that there's nothing bad with this server code per se but that the misconfiguration will come from the implementation of the main function in other files (e.g. alpn.c or certs.c).

jboeuf commented on test/core/bad_ssl/servers/cert.c:
```diff
@@ -0,0 +1,76 @@
+int main(int argc, char **argv) {
+  const char *addr = bad_ssl_addr(argc, argv);
+  grpc_ssl_pem_key_cert_pair pem_key_cert_pair;
+	grpc_server_credentials *ssl_creds;
```
clang-format

jboeuf commented on test/core/bad_ssl/servers/cert.c:
```diff
@@ -0,0 +1,76 @@
+#include "test/core/bad_ssl/server.h"
+#include "test/core/end2end/data/ssl_test_data.h"
+
+int main(int argc, char **argv) {
```
Intro comment:
regular client will connect to a server using untrusted certs.

jboeuf commented on test/core/bad_ssl/servers/alpn.c:
```diff
@@ -0,0 +1,81 @@
+
+static const char *const fake_versions[] = {"not-h2"};
+
+int grpc_chttp2_is_alpn_version_supported(const char *version, size_t size) {
```
Have a comment like: 
overriding function in src/core/transport/chttp2/alpn.c

jboeuf commented on test/core/bad_ssl/servers/alpn.c:
```diff
@@ -0,0 +1,81 @@
+  return 0;
+}
+
+size_t grpc_chttp2_num_alpn_versions(void) {
```
Have a comment like: 
overriding function from src/core/transport/chttp2/alpn.c

jboeuf commented on test/core/bad_ssl/servers/alpn.c:
```diff
@@ -0,0 +1,81 @@
+  return GPR_ARRAY_SIZE(fake_versions);
+}
+
+const char *grpc_chttp2_get_alpn_version_index(size_t i) {
```
Have a comment like: 
overriding function from src/core/transport/chttp2/alpn.c

jboeuf commented:
Thanks much for the PR! A few comments mostly about formatting and documentation.

ctiller commented:
Addressed review comments. I want to make one more tweak before finalizing though: at the moment although these tests do add coverage, they won't fail if the connection manages to get through. 

I'm going to address that with one more commit, and then ping this thread.

ctiller commented:
Ok, manually verified (by changing alpn.c) that this test now fails if we get a completed connection, as desired.

jboeuf commented:
LGTM. Merging.
